cli/src/main/java/com/devonfw/tools/ide/cli/CommandletRegistry.java

```diff
@@ -0,0 +1,184 @@
+package com.devonfw.tools.ide.cli;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.jline.console.ArgDesc;
+import org.jline.console.CmdDesc;
+import org.jline.console.CommandRegistry;
+import org.jline.reader.impl.completer.SystemCompleter;
+import org.jline.utils.AttributedString;
+
+import com.devonfw.tools.ide.commandlet.Commandlet;
+import com.devonfw.tools.ide.commandlet.ContextCommandlet;
+import com.devonfw.tools.ide.context.IdeContext;
+import com.devonfw.tools.ide.property.CommandletProperty;
+import com.devonfw.tools.ide.property.Property;
+
+/**
+ * Implements the {@link CommandRegistry} for jline3.
+ */
+public class CommandletRegistry implements CommandRegistry {
+
+  private final ContextCommandlet cmd;
+
+  private final IdeContext context;
+
+  private final Ide ide;
+
+  private final IdeCompleter ideCompleter;
+
+  private final Set<String> commandlets;
+
+  private final Map<String, String> aliasCommandlets;
+
+  public CommandletRegistry(ContextCommandlet cmd, Ide ide, IdeContext context) {
+
+    this.ideCompleter = new IdeCompleter(cmd, context);
+    this.cmd = cmd;
+    this.context = context;
+    this.ide = ide;
+
+    Set<String> commandlets = new HashSet<>();
+    Collection<Commandlet> commandletCollection = context.getCommandletManager().getCommandlets();
+
+    for (Commandlet commandlet : commandletCollection) {
+      commandlets.add(commandlet.getName());
+    }
+
+    // TODO: get correct aliases, uses command name keywords instead now
+    Map<String, String> aliasCommandlets = new HashMap<>();
+    Collection<Commandlet> aliasCommandletCollection = context.getCommandletManager().getCommandlets();
+
+    for (Commandlet commandlet : aliasCommandletCollection) {
+      aliasCommandlets.put(commandlet.getName(), commandlet.getKeyword());
+    }
+
+    this.commandlets = commandlets;
+    this.aliasCommandlets = aliasCommandlets;
+  }
+
+  @Override
+  public Set<String> commandNames() {
+
+    return commandlets;
+  }
+
+  @Override
+  public Map<String, String> commandAliases() {
+
+    return aliasCommandlets;
+  }
+
+  @Override
+  public List<String> commandInfo(String command) {
+
+    List<String> out = new ArrayList<>();
+    // TODO: take command and get help from Ide
+    Commandlet helpCommandlet = context.getCommandletManager().getCommandlet("help");
+    Property<?> property = new CommandletProperty(command, false, "");
+    // helpCommandlet.add(property);
+    helpCommandlet.run();
```

Comment:
If you really want to run the help for the commendlet then do this:
```
HelpCommandlet helpCommandlet = context.getCommandletManager().getCommandlet(HelpCommandlet.class);
helpCommandlet.commandlet.setValueAsString(command);
helpCommandlet.run();
```
However, this will print the help to the logger. While you could mock it and retrieve its output here, it would IMHO be better to refactor `HelpCommandlet` and `Ide` to allow better reuse here...

Reply:
Adjusted now.

Reply:
Thanks. Still it would make much sense to properly expose the logic to get help for better reuse than actually invoking the HelpCommandlet, but lets do this after the merge. There are tons of other things to rework. I do not like merging stuff that has lots of TODOs and things we already know that we need to rework, but in this case, we need to go forward.